pkg/controller/nodelink/nodelink_controller_test.go

```diff
@@ -0,0 +1,495 @@
+package nodelink
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"testing"
+	"time"
+
+	mapiv1beta1 "github.com/openshift/cluster-api/pkg/apis/machine/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/klog"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+func init() {
+	if err := mapiv1beta1.AddToScheme(scheme.Scheme); err != nil {
+		klog.Fatal(err)
+	}
+}
+
+const (
+	ownerControllerKind = "MachineSet"
+	namespace           = "openshift-machine-api"
+)
+
+var (
+	knownDate          = metav1.Time{Time: time.Date(1985, 06, 03, 0, 0, 0, 0, time.Local)}
+	fakeNodeIndexer    = make(map[string]corev1.Node)
+	fakeMachineIndexer = make(map[string]mapiv1beta1.Machine)
+)
+
+func node(name, providerID string, addresses []corev1.NodeAddress, taints []corev1.Taint) *corev1.Node {
+	node := &corev1.Node{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: metav1.NamespaceNone,
+		},
+		TypeMeta: metav1.TypeMeta{
+			Kind: "Node",
+		},
+		Status: corev1.NodeStatus{
+			Conditions: []corev1.NodeCondition{
+				{
+					Type:               corev1.NodeReady,
+					Status:             corev1.ConditionTrue,
+					LastTransitionTime: knownDate,
+				},
+			},
+		},
+	}
+
+	if providerID != "" {
+		node.Spec.ProviderID = providerID
+	}
+	if addresses != nil {
+		node.Status.Addresses = addresses
+	}
+	if taints != nil {
+		node.Spec.Taints = taints
+	}
+	return node
+}
+
+func machine(name, providerID string, addresses []corev1.NodeAddress, taints []corev1.Taint) *mapiv1beta1.Machine {
+	machine := &mapiv1beta1.Machine{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+			Labels: map[string]string{
+				"foo": "a",
+				"bar": "b",
+			},
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					Kind: ownerControllerKind,
+				},
+			},
+		},
+		TypeMeta: metav1.TypeMeta{
+			Kind: "Machine",
+		},
+		Spec: mapiv1beta1.MachineSpec{},
+	}
+
+	if providerID != "" {
+		machine.Spec.ProviderID = &providerID
+	}
+	if addresses != nil {
+		machine.Status.Addresses = addresses
+	}
+	if taints != nil {
+		machine.Spec.Taints = taints
+	}
+	return machine
+}
+
+func newFakeReconciler(client client.Client, machine *mapiv1beta1.Machine, node *corev1.Node) ReconcileNodeLink {
+	r := ReconcileNodeLink{
+		client: client,
+	}
+	r.listNodesByFieldFunc = func(_, value string) ([]corev1.Node, error) {
```

Comment:
Did you try to use the fake client itself? It should work as any other client where you just populated all node/machine objects on your own.

Reply:
yes, the fakeClient does not support fake indexers